core/src/main/kotlin/knote/PageManagerImpl.kt

```diff
@@ -262,12 +265,73 @@ internal class PageManagerImpl(
                 }
             }
         }
+        KNote.cancelOnShutDown(job)
+        watchPageJob = job
 
-        NotebookManagerImpl.logger.trace("started page watcher")
+        logger.trace("started page watcher")
     }
+    private var watchDataJob: Job? = null
 
+    private fun startDataWatcher() {
+        logger.debug("starting page watcher")
+        val job = watchActor(notebookScript.dataRoot.absoluteFile.toPath()) {
+//            var timeout: Job? = null
+            for (watchEvent in channel) {
+                val path = watchEvent.context()
+                val file = notebookScript.dataRoot.resolve(path.toFile()).absoluteFile
+                val event = watchEvent.kind()
+
+                logger.info("event: $path, ${event.name()}")
+
+                // TODO: readd editing timeout
+//                timeout?.cancel()
+//                timeout = launch {
+//                    delay(1000)
+
+                    when (event.name()) {
+                        "ENTRY_CREATE" -> {
+                            logger.debug("${watchEvent.context()} was created")
+                            pages.forEach { pageId, page ->
+                                if(path in page.fileInputs) {
+                                    executePage(pageId)
+                                }
+                            }
+                        }
+                        "ENTRY_MODIFY" -> {
+                            logger.debug("${watchEvent.context()} was modified")
+                            var executed = false
+                            pages.forEach { pageId, page ->
+                                if(path in page.fileInputs) {
+                                    executePage(pageId)
+                                    executed = true
+                                }
+                            }
+                            // ensure all pages have their results cached again
+                            if(executed) {
+                                notebookScript.pageFiles.forEach {
+                                    val id = it.name.substringBeforeLast(".page.kts")
+                                    val result = executePageCached(id)
+                                    logger.info("[$id] => $result")
+                                }
+                            }
+                        }
+                        "ENTRY_DELETE" -> {
+                            logger.debug("${watchEvent.context()} was deleted")
+                        }
+                        "OVERFLOW" -> logger.debug("${watchEvent.context()} overflow")
+                    }
+//                }
+            }
+        }
+        KNote.cancelOnShutDown(job)
+        watchPageJob = job
+
+        logger.trace("started data watcher")
+    }
     internal fun stopWatcher() {
-        watchJob?.cancel()
-        watchJob = null
+        watchPageJob?.cancel()
```

Comment:
I assume a data job has to do with imported data like csv, json, right? Maybe we should add a comment to define that at the top of wherever this is originally defined.